source/tests/unittests/TestHelixClass.cpp

```diff
@@ -47,3 +51,168 @@ TEMPLATE_LIST_TEST_CASE("Initialize_VP", "[helix-init]", HelixTypes) {
+  for (int i=0; i<3; i++) momentum[i] = momentum1[i] + momentum2[i];
+
+  const auto magField = 3.5;
+  const auto charge1 = -1.0;
+  const auto charge2 = 1.0;
+  const auto pt1 = std::hypot(momentum1[0], momentum1[1]);
+  const auto pt2 = std::hypot(momentum2[0], momentum2[1]);
+
+  const auto omega1 = c_helix * magField * charge1 / pt1;
+  const auto omega2 = c_helix * magField * charge2 / pt2;
+  const auto tanLambda1 = momentum1[2] / pt1;
+  const auto tanLambda2 = momentum2[2] / pt2;
+  const auto phi01 = atan2( momentum1[1],momentum1[0] );
+  const auto phi02 = atan2( momentum2[1],momentum2[0] );
+
+
+  FloatT vtx_momentum1[3];
+  FloatT vtx_momentum2[3];
+  FloatT vertex1[3];
+  FloatT vertex2[3];
+
+  SECTION("no reference point given and helices passing through zero") {
+    /*const*/ std::array<FloatT, 3> position = {0, 0, 0};
+
+    // passing through ref. point
+    const auto d0 = 1.0e-5;
+    const auto z0 = 1.0e-5;
+
+    helix1.Initialize_Canonical(phi01, d0, z0, omega1, tanLambda1, magField);
+    helix2.Initialize_Canonical(phi02, -d0, z0, omega2, tanLambda2, magField);
+
+    CHECK_THAT(helix1.getRadius(),
+                Catch::Matchers::WithinAbs(pt1 / (c_helix * magField), 0.1));
+    CHECK_THAT(helix2.getRadius(),
+                Catch::Matchers::WithinAbs(pt2 / (c_helix * magField), 0.1));
+
+
+    FloatT dist1 = helix1.getDistanceToHelix(&helix2, vertex1, vtx_momentum1);
+    FloatT dist2 = helix2.getDistanceToHelix(&helix1, vertex2, vtx_momentum2);
+
+    const auto eps = 0.1;
+
+    CHECK_THAT(0.0, Catch::Matchers::WithinAbs(dist1, eps));
+    CHECK_THAT(0.0, Catch::Matchers::WithinAbs(dist2, eps));
+
+    for (int i=0; i<3; i++) {
+      CHECK_THAT(position[i], Catch::Matchers::WithinAbs(vertex1[i], eps));
+      CHECK_THAT(position[i], Catch::Matchers::WithinAbs(vertex2[i], eps));
+
+      CHECK_THAT(momentum[i], Catch::Matchers::WithinAbs(vtx_momentum1[i], eps));
+      CHECK_THAT(momentum[i], Catch::Matchers::WithinAbs(vtx_momentum2[i], eps));
+    }
+  }
+
+  SECTION("input reference point and helices with small d0, z0") {
+    // "real-world" numbers (BSM event)
+    std::array<FloatT, 3> position = {242.5, -358.5, 171.6}; // true vtx
+    std::array<FloatT, 3> ref1 = {244.944, -359.603, 169.052};
+    std::array<FloatT, 3> ref2 = {249.609, -356.381, 198.409};
+
+    helix1.Initialize_Canonical(phi01, 0.103768967092, 1.39691245556, omega1,
+                                tanLambda1, magField, ref1.data());
+    helix2.Initialize_Canonical(phi02, 0.275440096855, -1.55072879791, omega2,
+                                tanLambda2, magField, ref2.data());
+
+    FloatT dist1 = helix1.getDistanceToHelix(&helix2, vertex1, vtx_momentum1);
+    FloatT dist2 = helix2.getDistanceToHelix(&helix1, vertex2, vtx_momentum2);
+
+    // distance should be small but nonzero
+    CHECK_THAT(0.0, Catch::Matchers::WithinAbs(dist1, 2));
+    CHECK_THAT(0.0, Catch::Matchers::WithinAbs(dist2, 2));
```

Comment:
Is 2 mm(?) a reasonable difference for it still to be considered non-zero? Or is this mainly to make these tests pass? Is it possible to calculate the actual expected distance in this case, and check that with a much smaller epsilon?

Reply:
It's partly to make the tests pass, but also, honestly, I don't see any way to calculate this, other than as currently implemented in `getDistanceToHelix()`. If the helices do not **cross** (which of course is the case in real life), the calculation becomes highly non-trivial, as one can see from complexity of the algorithm. So either we accept the current algorithm prediction as the "correct" and use it in the test, or assume the distance should be small and accept everything close to zero within some margin. 

I chose the latter option because: 1) I think 2 mm is a good precision for such a high track displacement 2) from my experience, when there is a bug in the algorithm, the calculated distance explodes, 3) even though I went through this code and I believe I understand how it works, it is still not mine and it might happen that I am missing something (as with the x_yCentre bug pointed out by @dudarboh)